platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/hashmap/ReusableBucketPool.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (C) 2021-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.merkledb.files.hashmap;
+
+import com.swirlds.virtualmap.VirtualKey;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * HalfDiskHashMap buckets are somewhat expensive resources. Every bucket has an
+ * underlying byte buffer to store bucket data and metadata, and the number of
+ * buckets is huge. This class provides a bucket pool, so buckets can be reused
+ * rather than created on every read / write call.
+ *
+ * <p>Bucket pool is accessed from multiple threads:
+ * <ul>
+ *     <li>Transaction thread, when a key path is loaded from HDHM as a part of
+ *     get or getForModify call</li>
+ *     <li>Lifecycle thread, when updated bucket is written to disk in the end
+ *     of HDHM flushing</li>
+ *     <li>HDHM background bucket reading threads</li>
+ *     <li>Warmup (aka prefetch) threads</li>
+ * </ul>
+ *
+ * <p>If buckets were created, updated, and then released (marked as available
+ * for other threads) on a single thread, this class would be as simple as a
+ * single {@link ThreadLocal} object. This is not the case, unfortunately. For
+ * example, when HDHM background reading threads read buckets from disk, buckets
+ * are requested from the pool by {@link BucketSerializer} as a part of data
+ * file collection read call. Then buckets are updated and put to a queue, which
+ * is processed on a different thread, virtual pipeline (aka lifecycle) thread.
+ * Only after that buckets can be reused. This is why the pool is implemented as
+ * an array of buckets with fast concurrent read/write access from multiple
+ * threads.
+ */
+public class ReusableBucketPool<K extends VirtualKey> {
+
+    /** Default number of reusable buckets in this pool */
+    private static final int DEFAULT_POOL_SIZE = 8192;
+
+    /** Pool size */
+    private final int poolSize;
+
+    /** Buckets */
+    private final AtomicReferenceArray<Bucket<K>> buckets;
```

Comment:
Just a thought.  We have three arrays that have to match on index.  Would it make sense to have a holder (e.g. `private static record BucketHolder<K>(Bucket<K> bucket, Lock lock, Condition condition)`) record and a single array?  Not sure if that entirely works with the ordering of events below, but might simplify a few things...

Reply:
Let me try it out

Reply:
Ah, it wouldn't work without significant changes... Buckets in this array are changed to null and back, while locks/conditions aren't. If `buckets` is changed to something like `AtomicReferenceArray<BucketHolder<K>>`, there will be no way to atomically getAndSet() elements.

Probably, instead of setting elements to nulls and back to real buckets, I would be able to have a final unmodifiable list of `BucketHolder<K>` records, with one of the fields to be AtomicBoolean, which indicates whether the bucket is currently in use or not. I'm not sure how ugly it would look, though.

Reply:
Thanks for checking.  Probably something for a future PR.

Reply:
This wouldn't work, either. With the current implementation, buckets order in the array is not fixed, buckets are shuffled in all different ways. To implement the idea above with atomic booleans, to release a bucket, it will have to be found in the array first, which is extra performance costs.